aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -218,32 +241,43 @@ private void uploadParts() {
 
     stagingFiles.stream()
         // do not upload the file currently being written
-        .filter(f -> closed || !f.equals(currentStagingFile))
+        .filter(f -> closed || !f.file().equals(currentStagingFile))
         // do not upload any files that have already been processed
-        .filter(Predicates.not(multiPartMap::containsKey))
-        .forEach(f -> {
+        .filter(Predicates.not(f -> multiPartMap.containsKey(f.file())))
+        .forEach(fileAndDigest -> {
+          File file = fileAndDigest.file();
           UploadPartRequest.Builder requestBuilder = UploadPartRequest.builder()
               .bucket(location.bucket())
               .key(location.key())
               .uploadId(multipartUploadId)
-              .partNumber(stagingFiles.indexOf(f) + 1)
-              .contentLength(f.length());
+              .partNumber(stagingFiles.indexOf(fileAndDigest) + 1)
+              .contentLength(file.length());
+
+          if (fileAndDigest.hasDigest()) {
+            requestBuilder.contentMD5(BinaryUtils.toBase64(fileAndDigest.getDigest()));
+          }
 
           S3RequestUtil.configureEncryption(awsProperties, requestBuilder);
 
           UploadPartRequest uploadRequest = requestBuilder.build();
 
           CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
               () -> {
-                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f));
+                UploadPartResponse response = null;
+                try {
+                  response = s3.uploadPart(uploadRequest, RequestBody.fromFile(file));
+                } catch (UncheckedIOException uncheckedIOException) {
+                  throw new UncheckedIOException(
```

Comment:
I am thinking stripping of message from original `UncheckedIOException` is probably OK as I did not see any helpful info in there, but then it's not guaranteed to remain the same. The security concern expressed by @jackye1995 on logging entire request (which I believe is fine as sensitive parts of requests are already redacted) is also an open conversation.

Reply:
What was the security concern? I don't think that this should be catching the exception and editing it, unless there's a strong reason. Since `UncheckedIOException` is a wrapper produced by other Iceberg code, it's very unlikely to add information that is a security concern.

Reply:
Security concern is mostly about s3 request info, previous thread is here https://github.com/apache/iceberg/pull/3813#discussion_r789155286

Reply:
Thanks for the info. I wouldn't worry about it. There is no need for error handling here so let's just let the exception propagate and we can worry about that concern in more appropriate places.